[Koh Kai Jie] iP

kohkaijie opened the pull request:
### CR7

> "I see myself as the best footballer in the world" - Cristiano Ronaldo

Ever feel like you had the potential to be a superstar like Cristiano Ronaldo, but just never had the confidence or motivation to do so? Do not worry, because CR7 is here! It helps you manage your tasks and it's,

- text-based
- **EASY** to learn
- **FAST** to use

All you need to do is,

1. download it from [here](https://www.youtube.com/watch?v=i8GTNSIBH58)
2. double-click it
3. add your tasks
4. let it manage your tasks for you! :kissing_closed_eyes:

And it is **FREE**!!!! SIUUUUUU

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon!)
- [ ] Reminders (coming soon!)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

kohkaijie committed:
level 0

kohkaijie committed:
level 1

kohkaijie committed:
level 2

kohkaijie committed:
level 3

kohkaijie committed:
level 4

kohkaijie committed:
level 5

kohkaijie committed:
level 6

kohkaijie committed:
A-TextUiTesting

kohkaijie committed:
level 7

kohkaijie committed:
level 8

kohkaijie committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2324S1/ip

kohkaijie committed:
More OOP

kohkaijie committed:
A-branch

kohkaijie committed:
level 9

kohkaijie committed:
coding standard

kohkaijie committed:
java docs

geraldngjx commented on src/main/java/functions/Parser.java:
```diff
@@ -0,0 +1,139 @@
+                        if (start == null || end == null) {
+                            return new ErrorCommand("PLease enter the date and time in the correct format. \n");
+                        }
+                        return new AddCommand(desc, start, end);
```
Good effort in abstracting out code into specific classes such as AddCommand, which makes the file length shorter and more readable!

geraldngjx replied:
Perhaps you can consider using switch statements to reduce the nesting of if-else statements?

geraldngjx commented on src/main/java/commands/Command.java:
```diff
@@ -0,0 +1,30 @@
+package commands;
+
+import functions.*;
```
Perhaps you can import the specific classes that you are using in this file instead of using the "*" notation?

geraldngjx commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,66 @@
+package commands;
+
+import functions.*;
```
Perhaps you want to import specific classes instead of using the "*" notation?

woojiahao replied:
Seconded! One of Intellij's auto-suggested actions is to expand all wildcard imports, you can try that!

geraldngjx commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,66 @@
+import functions.*;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import tasks.*;
```
As commented in other parts of the code, you might consider importing specific classes instead of using the "*" notation?

geraldngjx commented on src/main/java/commands/Command.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * The base abstract class for different types of commands.
+ */
+public abstract class Command {
```
Great job creating an abstract class Command to manage all existing commands which makes the code more readable!

geraldngjx reviewed (commented):
Your commitment to code quality shines through in this PR. It's evident that you've paid meticulous attention to detail, adhering to consistent naming conventions and maintaining a clean and readable codebase. Your choice of well-descriptive variable names significantly enhances code clarity. Kudos on your dedication to producing high-quality code!

However, one suggestion for further improvement is to consider importing specific files instead of using the "*" notation. While it's convenient, explicit imports can enhance code transparency and help prevent potential naming conflicts in the future. Keep up the good work!

LWZ19 commented on src/main/java/CR7.java:
```diff
@@ -0,0 +1,60 @@
+ */
+public class CR7 {
+    private Storage storage;
+    private Ui ui;
```
Perhaps you could put everything in a `cr7` package. :smiley:

LWZ19 commented on src/main/java/functions/Parser.java:
```diff
@@ -0,0 +1,139 @@
+                LocalDateTime d = LocalDateTime.parse(dateString, format.getFormatter());
+                return d;
+            } catch (DateTimeParseException e) {
+                // Try the next format
```
Nice! I like how you explicitly explain why this block is empty :blush:

LWZ19 commented on src/main/java/functions/Parser.java:
```diff
@@ -0,0 +1,139 @@
+            return formatter;
+        }
+    }
+
```
Great job! I like how you have an enum to open the possibility of different time formats and conventions.:thumbsup:
But, perhaps you could create a `DateTimeFormat` enum class?

LWZ19 commented on src/main/java/functions/Parser.java:
```diff
@@ -0,0 +1,139 @@
+                return new MarkCommand(s);
+            } else if (first.equals("unmark")) {
+                int s = Integer.valueOf(words[1]);
+                return new UnmarkCommand(s);
```
Perhaps you could put the result directly into the arguments? For example:
```suggestion
                return new UnmarkCommand(Integer.valueOf(words[1]));
```
I'm not so sure about the usage of `s` :confused:

LWZ19 commented on src/main/java/functions/Parser.java:
```diff
@@ -0,0 +1,139 @@
+            }
+        }
+    }
+}
```
I like how you abstract all the user commands into their own classes which made the code clean and easy to understand, but perhaps you could make it even cleaner through switch-case statements?

LWZ19 commented on src/main/java/functions/Parser.java:
```diff
@@ -0,0 +1,139 @@
+package functions;
+
+import commands.*;
```
I noticed the same issue of '`*`' in several other places too. :worried:

LWZ19 commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,66 @@
+     * @param ui      The user interface for displaying messages.
+     * @param storage The storage interface for saving data.
+     * @throws IOException If an I/O error occurs while interacting with storage.
+     */
```
Perhaps you could add the `@Override` annotation for abstract methods? It can improve readability and help you debug if the method is not overridden properly. :smile:

LWZ19 commented on src/main/java/CR7.java:
```diff
@@ -0,0 +1,60 @@
+import commands.*;
+import functions.*;
```
Perhaps you could reduce the usage of wildcard('`*`') when importing classes, it's better to have multiple-line import statements. You could configure some IDE settings to allow this and let it write the import for you. :blush:

LWZ19 commented on src/main/java/CR7.java:
```diff
@@ -0,0 +1,60 @@
+    public static void main(String[] args) {
+        new CR7("src/main/data/CR7.txt").run();
+    }
+}
```
I like how you took the effort :thumbsup: to write comments in subtle places and JavaDoc for all the methods, which made your code very easy to understand!

LWZ19 commented on src/main/java/CR7.java:
```diff
@@ -0,0 +1,60 @@
+/**
+ * The main class for running the CR7 task management application.
+ */
+public class CR7 {
```
I like that you created your own class CR7.

LWZ19 reviewed (commented):
LGTM. Some suggestions are that you could put everything inside a package, and reduce the usage of '*' in imports. 

Other than those you are on par with the coding standard, good naming, layout, statements and comments. You’re doing an excellent job, keep it up! :tada:

woojiahao commented on src/main/data/CR7.txt:
Perhaps it's better to add the data file to your `.gitignore` so that when others clone your repository, they start from scratch instead

woojiahao commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,66 @@
+     */
+    public AddCommand(String desc) {
+        this.desc = desc;
+        this.first = null;
+        this.second = null;
```
Perhaps you do not need the `this` call for `first` and `second`. This is inline with the advanced coding standard of avoiding unnecessary `this` calls

woojiahao commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,66 @@
+     *
+     * @param desc The description of the task to be added.
+     */
+    public AddCommand(String desc) {
```
Perhaps you can explore using the `this` constructor instead? So pass everything to the `AddCommand` constructor with 3 parameters instead

woojiahao commented on src/main/java/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    /**
+     * Executes the DeleteCommand to delete a task, display a message, and save changes.
```
Great that your JavaDocs are descriptive!

woojiahao commented on src/main/java/commands/SearchCommand.java:
```diff
@@ -0,0 +1,38 @@
+    /**
+     * The keyword to search for in task descriptions.
+     */
+    String searchWord;
```
Perhaps you can provide a proper access modifier for this field if it is not going to be publicly accessed outside of the class

woojiahao commented on src/main/java/functions/Parser.java:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Enumeration representing different date and time formats for parsing.
+     */
+    enum DateTimeFormat {
```
Interesting use of enums to handle the date formats!

woojiahao commented on src/main/java/functions/Storage.java:
```diff
@@ -0,0 +1,83 @@
+    public void saveFiles(ArrayList<Task> tasks) throws IOException {
+        int i = tasks.size();
+        String taskList = "";
+        for (int j = 0; j < i; j++) {
+            if (j == i - 1) {
+                taskList += tasks.get(j).toString();
+            } else {
+                taskList += tasks.get(j).toString() + System.lineSeparator();
+            }
+        }
```
Perhaps you can explore using `String.join` instead of an explicit `for` loop, minor nit though!

woojiahao commented on src/main/java/functions/TaskList.java:
```diff
@@ -0,0 +1,170 @@
+    /**
+     * Constructs an empty TaskList object.
+     */
+    public TaskList() {
+        this.tasks = new ArrayList<>();
+    }
```
Perhaps you could explore calling the other constructor through `this` and then provide an empty array list that way

woojiahao commented on src/main/java/functions/TaskList.java:
```diff
@@ -0,0 +1,170 @@
+     */
+    private void loadTask(String input) {
+        if (input.startsWith("[D]")) {
+            int y = input.indexOf("(by: ");
```
Perhaps the variable name could be more descriptive? If there is a name conflict, you can consider `byIdx` instead

woojiahao commented on src/main/java/functions/Ui.java:
```diff
@@ -0,0 +1,143 @@
+/**
+ * Manages user interface interactions and displays messages to the user.
+ */
+public class Ui {
```
Great you have a utility class to handle Ui printing! Very clean

woojiahao commented on src/main/java/tasks/Event.java:
```diff
@@ -0,0 +1,46 @@
+    public String toString() {
+        String newStart = start.format(DateTimeFormatter.ofPattern("MMM d yyyy hh:mm a"));
+        String newEnd = end.format(DateTimeFormatter.ofPattern("MMM d yyyy hh:mm a"));
+        return "[E]" + super.toString() + " (from: " + newStart + " to: " + newEnd + ")";
```
Perhaps you can use `String.format` instead of String concatenation

woojiahao reviewed (commented):
Overall great work! There are minor nits that have been mentioned that could be improved but keep it up!

kohkaijie committed:
Add GUI for chatbot

kohkaijie committed:
Add two J-Unit tests

kohkaijie committed:
Assertions: add Assertions to code

There may be potential bugs that violate how the code should behave.

Use of assert feature to document important assumptions that should hold
at various points in the code.

We ensure that during runtime that bugs are discovered and execution
is terminated early.

kohkaijie committed:
Merge pull request #1 from kohkaijie/branch-A-Assertions

Assertions: add Assertions to code

kohkaijie committed:
FixedDuration task class: Add support for FixedDuration task type

CR7 did not provide support for tasks with a fixed duration but no
start or end time.

Adding the FixedDuration task class allows users to manage such tasks.

Let's add the FixedDuration class as a subclass of the Task class.

Using inheritance as Fixed Duration tasks are a specific type of task
with similar characteristics and hence suitable as a subclass.

kohkaijie committed:
A-CheckStyle

kohkaijie committed:
Improve coding standard of program

The code has many areas which violate the coding standard such as
heavily nested loops and overly long methods which affect the
readability of the code.

Improving the coding standard by rewriting overly long methods to
shorter ones as well as avoiding nested loops can prevent issues in
the future where the code is hard to understand and read.

Let's improve the coding standard by rewriting parts of the code
which violate the coding standard.

kohkaijie committed:
Merge pull request #3 from kohkaijie/branch-A-CodingStandard

Improve coding standard of program

kohkaijie committed:
Update README.md

kohkaijie committed:
Update README.md

kohkaijie committed:
Update README.md

kohkaijie committed:
Update README.md

kohkaijie committed:
Update README.md

kohkaijie committed:
Cleaning up exception handling

kohkaijie committed:
Merge branch 'master' of https://github.com/kohkaijie/ip

kohkaijie committed:
Added Ui.png

kohkaijie committed:
Update README.md

kohkaijie committed:
Update README.md

kohkaijie committed:
Updated build and Ui.png

kohkaijie committed:
Merge branch 'master' of https://github.com/kohkaijie/ip

kohkaijie committed:
Update README.md

kohkaijie committed:
minor bug fix

kohkaijie committed:
Merge branch 'master' of https://github.com/kohkaijie/ip

kohkaijie committed:
Minor bug fixes

kohkaijie committed:
minor big fixes

kohkaijie committed:
Code quality and minor bug fixed

kohkaijie committed:
minor bug fixes
